src/tools/PublicAnalyzers/AvaloniaPropertyAnalyzer.cs

```diff
@@ -0,0 +1,463 @@
+        Category,
+        DiagnosticSeverity.Warning,
+        isEnabledByDefault: true,
+        "AvaloniaProperty objects have static lifetimes and should be created only once. To ensure this, only call Register or AddOwner in static constructors or static initializers.");
+
+    private static readonly DiagnosticDescriptor PropertyOwnedByGenericType = new(
+        "AVP1002",
+        "AvaloniaProperty objects should not be owned by a generic type",
+        "Inadvisable registration: Generic types cannot be referenced from XAML. Create a non-generic type to be the owner of this AvaloniaProperty.",
+        Category,
+        DiagnosticSeverity.Warning,
+        isEnabledByDefault: true,
+        "It is sometimes necessary to refer to an AvaloniaProperty in XAML by providing its class name. This cannot be achieved if property's owner is a generic type." +
+        " Additionally, a new AvaloniaProperty object will be generated each time a new version of the the generic owner type is constructed, which may be unexpected.");
+
+    private static readonly DiagnosticDescriptor OwnerDoesNotMatchOuterType = new(
+        "AVP1010",
+        "AvaloniaProperty objects should be owned by the type in which they are stored",
+        "Type mismatch: AvaloniaProperty owner is {0}, which is not the containing type",
+        Category,
+        DiagnosticSeverity.Warning,
+        isEnabledByDefault: true,
+        "The owner of an AvaloniaProperty should generally be the containing type. This ensures that the property can be used as expected in XAML.",
+        TypeMismatchTag);
+
+    private static readonly DiagnosticDescriptor UnexpectedPropertyAccess = new(
+        "AVP1011",
+        "An AvaloniaObject should own each AvaloniaProperty it reads or writes on itself",
+        "Unexpected property use: {0} is neither owned by nor attached to {1}",
+        Category,
+        DiagnosticSeverity.Warning,
+        isEnabledByDefault: true,
+        "It is possible to use any AvaloniaProperty with any AvaloniaObject. However, each AvaloniaProperty an object uses on itself should be either owned by that object, or attached to that object.",
+        InappropriateReadWriteTag);
+
+    private static readonly DiagnosticDescriptor SettingOwnStyledPropertyValue = new(
+        "AVP1012",
+        "An AvaloniaObject should not set its own StyledProperty or AttachedProperty values",
+        "Inappropriate assignment: An AvaloniaObject should not set its own StyledProperty or AttachedProperty values",
+        Category,
+        DiagnosticSeverity.Warning,
+        isEnabledByDefault: true,
+        "Setting a StyledProperty or AttachedProperty directly will overwrite values from lower-priority sources, such as styles and templates. This should only be done by control consumers," +
+        "not the control itself. Constructor parameters and assignments within UserControl or TopLevel types are exempt from this diagnostic.",
```

Comment:
This is one is great! It's an easy one to forget in Avalonia especially with current code bases without SetCurrentValue(). Perhaps the wording can instruct devs they should call that method though.

Reply:
That's something to change when that method arrives in master. :)